R/webApiDataSources.R

```diff
@@ -0,0 +1,118 @@
+# @file dataSources
+#
+# Copyright 2019 Observational Health Data Sciences and Informatics
+#
+# This file is part of StudyManagement
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#' Get the results from individual Achilles reports by data Source definition expression
+#'
+#' @details
+#' Obtain the JSON expression from WebAPI for a given dataSource
+#'
+#' @param baseUrl        The base URL for the WebApi instance, for example:
+#'                       "http://server.org:80/WebAPI".
+#' @param sourceKey      The source key for a CDM instance in WebAPI, as defined in the
+#'                       Configuration page
+#' @param reportType     The precomputed report type in Atlas (Achilles report e.g. dashboard, datadensity, person etc)
+#' @return
+#' A list of object with pre computed Achilles results for the sourceKey and Achilles report
+#'
+#' @examples
+#' \dontrun{
+#' # This will obtain pre-computed summary regarding dataSource
+#'
+#' getDataSourceReportForSourceKey(baseUrl = "http://server.org:80/WebAPI", dataSource = 'dataSource', reportType = 'dashboard')
```

Comment:
What is the 'data source report'? How would one use the JSON that comes out of this function?

Reply:
I don't understand. How would you 'recreate another Achilles like instance'? What does that even mean?

Reply:
There are many reasons to use the raw Json. 

1. having the raw and full json archived in the study folder, allows us to go back and restore it in another instance of Achilles. Why is this important: lets say two years from now, we were asked why certain design choices were made (or why alternatives were not considered). If the answer to this question was not already documented, then we could go to the restored achilles
2. the parsed data objects derived by parsing JSON into R-friendly objects like data frames, are based on parsing logic. Our parsing logic is not 100% reversible to JSON i.e. we cannot create an Achilles instance from the parsed data.

Reply:
I'm aware of the need to be reproducible. But I would still like to know how this would be done. Do you mean there's another WebApi function that takes this JSON as input and creates a new Achilles instance? I thought Achilles nowadays is stored as database tables, not as JSON objects?

The reason I'm pushing back on this is because I don't want sham reproducibility, where we store a massive text blob and say we're reproducible, while there's no way to use that blob. 

I'm also not sure whether storing all Achilles data study-side is the solution. We don't do the same thing for CDM data: If I run a study, I store a reference to the exact CDM database I used, and rely on our DB admins to make sure in the future we can retrieve that CDM database using the reference. I don't store a full copy of the CDM database with my study. Why use different behavior here?

Reply:
makes sense. we could revisit this thought when there is a need to create a new full/partial Achilles (like) instance. For now, i agree with you - lets not call ourselves reproducible by making copies of blobs of text.!